### PR TITLE
KAFKA-13279: KRaft: add CreateTopicsPolicy + AlterConfigsPolicy

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

**Comment by dengziming** [comment]
Should we also change the integration test `CreateTopicsRequestWithPolicyTest` to support KRaft mode? the change LGTM.

File: clients/src/main/java/org/apache/kafka/server/policy/CreateTopicPolicy.java [hunk 1]
```diff
@@ -104,6 +105,24 @@ public Short replicationFactor() {
...
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            RequestMetadata other = (RequestMetadata) o;
+            return topic.equals(other.topic) &&
+                Objects.equals(numPartitions, other.numPartitions) &&
+                Objects.equals(replicationFactor, other.replicationFactor) &&
+                Objects.equals(replicasAssignments, other.replicasAssignments) &&
+                configs.equals(other.configs);
```
Comment (niket-goel) [inline clients/src/main/java/org/apache/kafka/server/policy/CreateTopicPolicy.java:123]: Questions for my own understanding:
Is there a particular reason we are not using the Object.equals method to compare, which is also a map?

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 2]
```diff
@@ -465,6 +485,13 @@ private ApiError createTopic(CreatableTopic topic,
...
             }
+            ApiError error = maybeCheckCreateTopicPolicy(() -> {
+                Map<String, String> configs = new HashMap<>();
+                topic.configs().forEach(config -> configs.put(config.name(), config.value()));
+                return new CreateTopicPolicy.RequestMetadata(
+                    topic.name(), numPartitions, replicationFactor, null, configs);
+            });
+            if (error.isFailure()) return error;
         }
         Uuid topicId = Uuid.randomUuid();
```
Comment (niket-goel) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:504]: nit: 
Would it be more readable to combine the two calls to `maybeCheckCreateTopicPolicy` and place it after the if/else blocks?

**Review by niket-goel — COMMENTED** [review]
Change looks good to me. Have a minor comment.

File: clients/src/main/java/org/apache/kafka/server/policy/CreateTopicPolicy.java (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/server/policy/CreateTopicPolicy.java:123]: `configs` can not be null. `numPartitions`, `replicationFactor` and `replicasAssignments` can be null.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 2 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:504]: The calls are doing different things based on whether we are using manual partition assignment or automatic. So they cannot be combined in a reasonable way.

**Comment by cmccabe** [comment]
> Should we also change the integration test CreateTopicsRequestWithPolicyTest to support KRaft mode? the change LGTM.

I do want to do this but I think we'll need a separate PR for it. We have to restructure the base classes for the test a bit due to the complex web of inheritance it's embedded in.

**Review by mumrah — APPROVED** [review]
Thanks @cmccabe, this LGTM. This would be a good candidate for converting an integration test to KRaft (i.e., AdminClientWithPoliciesIntegrationTest).

**Pull request merged**